streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java

```diff
@@ -184,6 +247,75 @@ public void process(final Record<KIn, VIn> record) {
                     droppedRecordsSensor.record();
                 }
             }
+
+            maybeMeasureLatency(() -> tryEmitFinalResult(record, closeTime), time, emitFinalLatencySensor);
+        }
+
+        private void tryEmitFinalResult(final Record<KIn, VIn> record, final long closeTime) {
+            if (emitStrategy.type() != StrategyType.ON_WINDOW_CLOSE) {
+                return;
+            }
+
+            final long now = internalProcessorContext.currentSystemTimeMs();
+            // Throttle emit frequency
+            if (now < timeTracker.nextTimeToEmit) {
+                return;
+            }
+
+            // Schedule next emit time based on now to avoid the case that if system time jumps a lot,
+            // this can be triggered everytime
+            timeTracker.nextTimeToEmit = now;
+            timeTracker.advanceNextTimeToEmit();
+
+            // Close time does not progress
+            if (lastEmitCloseTime != ConsumerRecord.NO_TIMESTAMP && lastEmitCloseTime >= closeTime) {
+                return;
+            }
+
+            final long emitWindowStart = closeTime - windows.size();
```

Comment:
It took we some time to figure out what this variable really is. It's the upper bound of the emit range (based on window-start) right?

Maybe we can find a better name? Maybe `emitRangeUpperBoundInclusive` ?

Similarly it might be helpful to introduce an `emitRangeLowerBoundInclusive` variable?

Reply:
Yeah. Will rename. Naming is hard...